handlers/update-supporter-plus-amount/src/updateSupporterPlusAmount.ts

```diff
@@ -0,0 +1,219 @@
+import { getAccount } from '@modules/zuora/getAccount';
+import { getSubscription } from '@modules/zuora/getSubscription';
+import type { ZuoraClient } from '@modules/zuora/zuoraClient';
+import type {
+	RatePlan,
+	RatePlanCharge,
+	ZuoraSubscription,
+} from '@modules/zuora/zuoraSchemas';
+import type { Dayjs } from 'dayjs';
+import dayjs from 'dayjs';
+import type { EmailFields } from './sendEmail';
+import { supporterPlusAmountBands } from './supporterPlusAmountBands';
+import { doUpdate } from './zuoraApi';
+
+type UpdatablePlans =
+	| 'Annual'
+	| 'Monthly'
+	| 'V1DeprecatedAnnual'
+	| 'V1DeprecatedMonthly';
+
+export type SupporterPlusData = {
+	ratePlan: RatePlan;
+	productRatePlan: ProductRatePlan<'SupporterPlus', UpdatablePlans>;
+	contributionCharge: RatePlanCharge;
+	planHasSeparateContributionCharge: boolean;
+};
+
+export const getSupporterPlusPlans = (
+	productCatalog: ProductCatalog,
+	ratePlans: RatePlan[],
+): SupporterPlusData => {
+	const supporterPlusProductData = {
+		[productCatalog.SupporterPlus.ratePlans.Monthly.id]: {
+			productRatePlan: productCatalog.SupporterPlus.ratePlans.Monthly,
+			charge:
+				productCatalog.SupporterPlus.ratePlans.Monthly.charges.Contribution,
+			planHasSeparateContributionCharge: true,
+		},
+		[productCatalog.SupporterPlus.ratePlans.Annual.id]: {
+			productRatePlan: productCatalog.SupporterPlus.ratePlans.Annual,
+			charge:
+				productCatalog.SupporterPlus.ratePlans.Annual.charges.Contribution,
+			planHasSeparateContributionCharge: true,
+		},
+		[productCatalog.SupporterPlus.ratePlans.V1DeprecatedMonthly.id]: {
+			productRatePlan:
+				productCatalog.SupporterPlus.ratePlans.V1DeprecatedMonthly,
+			charge:
+				productCatalog.SupporterPlus.ratePlans.V1DeprecatedMonthly.charges
+					.Subscription,
+			planHasSeparateContributionCharge: false,
+		},
+		[productCatalog.SupporterPlus.ratePlans.V1DeprecatedAnnual.id]: {
+			productRatePlan:
+				productCatalog.SupporterPlus.ratePlans.V1DeprecatedAnnual,
+			charge:
+				productCatalog.SupporterPlus.ratePlans.V1DeprecatedAnnual.charges
+					.Subscription,
+			planHasSeparateContributionCharge: false,
+		},
+	};
+
+	const productRatePlans = ratePlans
+		.filter((ratePlan) => ratePlan.lastChangeType !== 'Remove')
+		.reduce((acc: SupporterPlusData[], ratePlan: RatePlan) => {
+			const productData = supporterPlusProductData[ratePlan.productRatePlanId];
+			if (productData !== undefined) {
+				const { productRatePlan, planHasSeparateContributionCharge } =
+					productData;
+				const contributionCharge = checkDefined(
+					ratePlan.ratePlanCharges.find(
+						(charge) =>
+							charge.productRatePlanChargeId === productData.charge.id,
+					),
+					'Contribution charge not found in rate plan',
+				);
+				return [
+					...acc,
+					{
+						ratePlan,
+						productRatePlan,
+						contributionCharge,
+						planHasSeparateContributionCharge,
+					},
+				];
+			}
+			return acc;
+		}, []);
```

Comment:
having looked at the code, reduce is overkill anyway, all you are doing is mapping and filtering at the same time, you're not using the preceding values in any way.
If there's no flatMap or similar in typescript it could be done with filter and map
```suggestion
    .filter((ratePlan) => supporterPlusProductData[ratePlan.productRatePlanId] !== undefined)
		.map((acc: SupporterPlusData[], ratePlan: RatePlan) => {
			const productData = supporterPlusProductData[ratePlan.productRatePlanId];
			const contributionCharge = checkDefined(
				ratePlan.ratePlanCharges.find(
					(charge) =>
						charge.productRatePlanChargeId === productData.charge.id,
				),
				'Contribution charge not found in rate plan',
			);
			const { productRatePlan, planHasSeparateContributionCharge } = productData;
			return {
				ratePlan,
				productRatePlan,
				contributionCharge,
				planHasSeparateContributionCharge,
			};
		}, []);
```

Reply:
I've refactored to a filter and map